testwise deletion

jdramsey opened the pull request:
Added testwise deletion for all of the tests and scores that we use.

jdramsey commented:
Here's a simple session where you can try all of the combinations.

[testwise.deletion.practice.zip](https://github.com/cmu-phil/tetrad/files/5411248/testwise.deletion.practice.zip)

The column s are discrete, continuous, mixed, the rows are tests, scores. For tests, you can use PCall; for scores you can use FGES. Each of these should work with any arbitrary test or score (though some are very slow). When I say "work" I mean they shouldn't yield exceptions when missing values are added to the data. The "data" box in each case injects missing data. The lower right search box in each searches over the data with missing values injected.

Every algorithm should now either do testwise deletion automatically or else tell you to remove or impute missing data.

cg09 commented:
What version of TETRAD do I run it on?

On Tue, Oct 20, 2020 at 4:07 PM Joseph Ramsey <notifications@github.com>
wrote:

> Here's a simple session where you can try all of the combinations.
>
> testwise.deletion.practice.zip
> <https://github.com/cmu-phil/tetrad/files/5411248/testwise.deletion.practice.zip>
>
> The column s are discrete, continuous, mixed, the rows are tests, scores.
> For tests, you can use PCall; for scores you can use FGES. Each of these
> should work with any arbitrary test or score (though some are very slow).
> When I say "work" I mean they shouldn't yield exceptions when missing
> values are added to the data. The "data" box in each case injects missing
> data. The lower right search box in each searches over the data with
> missing values injected.
>
> Every algorithm should now either do testwise deletion automatically or
> else tell you to remove or impute missing data.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1307#issuecomment-713108558>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AD4Y3OJ3NSCVXXWZFLOHJEDSLXUWTANCNFSM4SYZGLOQ>
> .
>

kvb2univpitt commented:
@jdramsey Maven tests failed in tetrad-lib.  I got the following error:

```
java.lang.AssertionError: expected:<0.88> but was:<0.5738747071856971>
	at org.junit.Assert.fail(Assert.java:88)
	at org.junit.Assert.failNotEquals(Assert.java:834)
	at org.junit.Assert.assertEquals(Assert.java:553)
	at org.junit.Assert.assertEquals(Assert.java:683)
	at edu.cmu.tetrad.test.TestLingamPattern.test1(TestLingamPattern.java:85)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)

```
Also, I get a bunch of **NullPointerException** error:
```
java.lang.NullPointerException
	at edu.cmu.tetrad.search.IndTestFisherZ.getRows(IndTestFisherZ.java:485)
	at edu.cmu.tetrad.search.IndTestFisherZ.getPValue(IndTestFisherZ.java:232)
	at edu.cmu.tetrad.search.IndTestFisherZ.isIndependent(IndTestFisherZ.java:196)
	at edu.cmu.tetrad.search.FasStable.searchAtDepth0(FasStable.java:284)
	at edu.cmu.tetrad.search.FasStable.search(FasStable.java:164)
	at edu.cmu.tetrad.algcomparison.algorithm.oracle.pattern.PcFges.search(PcFges.java:47)
	at edu.cmu.tetrad.test.TestFges.testFromData(TestFges.java:872)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
java.lang.NullPointerException
```

The test never ends.

jdramsey commented:
Will fix. I do not think it will be difficult.

On Wed, Oct 21, 2020 at 9:23 PM Kevin Bui <notifications@github.com> wrote:

> @jdramsey <https://github.com/jdramsey> Maven tests failed in tetrad-lib.
> I got the following error:
>
> java.lang.AssertionError: expected:<0.88> but was:<0.5738747071856971>
> 	at org.junit.Assert.fail(Assert.java:88)
> 	at org.junit.Assert.failNotEquals(Assert.java:834)
> 	at org.junit.Assert.assertEquals(Assert.java:553)
> 	at org.junit.Assert.assertEquals(Assert.java:683)
> 	at edu.cmu.tetrad.test.TestLingamPattern.test1(TestLingamPattern.java:85)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
> 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
> 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
> 	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
> 	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
> 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
> 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
>
>
> Also, I get a bunch of *NullPointerException* error:
>
> java.lang.NullPointerException
> 	at edu.cmu.tetrad.search.IndTestFisherZ.getRows(IndTestFisherZ.java:485)
> 	at edu.cmu.tetrad.search.IndTestFisherZ.getPValue(IndTestFisherZ.java:232)
> 	at edu.cmu.tetrad.search.IndTestFisherZ.isIndependent(IndTestFisherZ.java:196)
> 	at edu.cmu.tetrad.search.FasStable.searchAtDepth0(FasStable.java:284)
> 	at edu.cmu.tetrad.search.FasStable.search(FasStable.java:164)
> 	at edu.cmu.tetrad.algcomparison.algorithm.oracle.pattern.PcFges.search(PcFges.java:47)
> 	at edu.cmu.tetrad.test.TestFges.testFromData(TestFges.java:872)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
> 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
> 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
> 	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
> 	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
> 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
> 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
> java.lang.NullPointerException
>
> The test never ends.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1307#issuecomment-714135435>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACLFSR7MITKXZNCXAYPI6ZLSL6CSJANCNFSM4SYZGLOQ>
> .
>


-- 
Joseph D. Ramsey
Special Faculty and Director of Research Computing
Department of Philosophy
135 Baker Hall
Carnegie Mellon University
Pittsburgh, PA 15213

jsph.ramsey@gmail.com
Office: (412) 268-8063
http://www.andrew.cmu.edu/user/jdramsey

jdramsey commented:
The loop was for an algorithm, PcFges, that was experimental many years ago and which we do not support. Removed the test.

There was a test that was hanging, for DM, reason being that several of the tests in that file use a sample size of 100,000, which is too much when you're checking for testwise NaN rows. If you give the algorithms covariance matrices with no NaN's in the calculation this problem goes away.

There were some cases where in my zeal to do testwise deletion I didn't go back and make sure the algorithms worked with covariance matrices and correlation matrices as inputs, and the test correctly caught these problems. I fixed a couple; I'll go through this morning and fix that issue more systematically.

Tests pass. I'll commit this work and then go back and work on the covariance issue more generally.

jdramsey commented:
The DM issue suggests that an improvement may be needed for testwise deletion to speed it up. A good option, I think, is to incorporate Kun's R variables, which are binary variables indicating where data are missing for each variable. I think it will be faster to loop over the R variables than over the original data, looking for NaN's. I'll try this. I don't know if I can get the whole way to 100,000 samples or not quickly,  but it's worth a shot.

jdramsey commented:
I think the covariance matrix issue is good. This only affects the continuous case. For the mixed case, we need to have a full dataset, as do we for KCI and CCI. For the discrete case, we need the full dataset. That leaves Fisher Z and SEM BIC, both of which I just fixed. If you want to use a covariance matrix, you need to use one of those.

All fixes done.

jdramsey commented:
I think the speedup of testwise deletion can wait until after the current round of fixes are completely done.

kvb2univpitt commented:
@jdramsey Please let me know when it's ready for me to test again.

jdramsey commented:
Go for it. I'm just adjusting a thing or two, but I'm done.

On Thu, Oct 22, 2020 at 10:40 AM Kevin Bui <notifications@github.com> wrote:

> @jdramsey <https://github.com/jdramsey> Please let me know when it's
> ready for me to test again.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1307#issuecomment-714541326>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACLFSR3RMUIGIUWBBVDTJ43SMA74FANCNFSM4SYZGLOQ>
> .
>


-- 
Joseph D. Ramsey
Special Faculty and Director of Research Computing
Department of Philosophy
135 Baker Hall
Carnegie Mellon University
Pittsburgh, PA 15213

jsph.ramsey@gmail.com
Office: (412) 268-8063
http://www.andrew.cmu.edu/user/jdramsey

kvb2univpitt commented:
@jdramsey I still got an error from Junit test:

```
Tests run: 6, Failures: 0, Errors: 1, Skipped: 4, Time elapsed: 2.884 sec <<< FAILURE!
testFGESc(edu.cmu.tetrad.test.TestGeneralResamplingTest)  Time elapsed: 1.815 sec  <<< ERROR!
org.apache.commons.math3.linear.SingularMatrixException: matrix is singular
	at org.apache.commons.math3.linear.LUDecomposition$Solver.solve(LUDecomposition.java:334)
	at org.apache.commons.math3.linear.LUDecomposition$Solver.getInverse(LUDecomposition.java:387)
	at edu.cmu.tetrad.util.Matrix.inverse(Matrix.java:181)
	at edu.cmu.tetrad.util.StatUtils.partialCorrelationPrecisionMatrix(StatUtils.java:1752)
	at edu.cmu.tetrad.util.StatUtils.partialCorrelation(StatUtils.java:1748)
	at edu.cmu.tetrad.search.SemBicScore.partialCorrelation(SemBicScore.java:486)
	at edu.cmu.tetrad.search.SemBicScore.localScoreDiff(SemBicScore.java:130)
	at edu.cmu.tetrad.search.Fges.scoreGraphChange(Fges.java:2027)
	at edu.cmu.tetrad.search.Fges.deleteEval(Fges.java:1525)
	at edu.cmu.tetrad.search.Fges.calculateArrowsBackward(Fges.java:1369)
	at edu.cmu.tetrad.search.Fges.initializeArrowsBackward(Fges.java:1106)
	at edu.cmu.tetrad.search.Fges.bes(Fges.java:1022)
	at edu.cmu.tetrad.search.Fges.search(Fges.java:265)
	at edu.cmu.tetrad.algcomparison.algorithm.oracle.pattern.Fges.search(Fges.java:90)
	at edu.pitt.dbmi.algo.resampling.task.GeneralResamplingSearchRunnable.run(GeneralResamplingSearchRunnable.java:133)
	at edu.pitt.dbmi.algo.resampling.GeneralResamplingSearch.search(GeneralResamplingSearch.java:360)
	at edu.pitt.dbmi.algo.resampling.GeneralResamplingTest.search(GeneralResamplingTest.java:199)
	at edu.cmu.tetrad.test.TestGeneralResamplingTest.testFGESc(TestGeneralResamplingTest.java:116)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
```

jdramsey commented:
I'm not seeing this when I run my tests in IntelliJ. Let me do a mav clean
test from the command line.

On Sat, Oct 24, 2020 at 11:30 AM Kevin Bui <notifications@github.com> wrote:

> @jdramsey <https://github.com/jdramsey> I still got an error from Junit
> test:
>
> Tests run: 6, Failures: 0, Errors: 1, Skipped: 4, Time elapsed: 2.884 sec <<< FAILURE!
> testFGESc(edu.cmu.tetrad.test.TestGeneralResamplingTest)  Time elapsed: 1.815 sec  <<< ERROR!
> org.apache.commons.math3.linear.SingularMatrixException: matrix is singular
> 	at org.apache.commons.math3.linear.LUDecomposition$Solver.solve(LUDecomposition.java:334)
> 	at org.apache.commons.math3.linear.LUDecomposition$Solver.getInverse(LUDecomposition.java:387)
> 	at edu.cmu.tetrad.util.Matrix.inverse(Matrix.java:181)
> 	at edu.cmu.tetrad.util.StatUtils.partialCorrelationPrecisionMatrix(StatUtils.java:1752)
> 	at edu.cmu.tetrad.util.StatUtils.partialCorrelation(StatUtils.java:1748)
> 	at edu.cmu.tetrad.search.SemBicScore.partialCorrelation(SemBicScore.java:486)
> 	at edu.cmu.tetrad.search.SemBicScore.localScoreDiff(SemBicScore.java:130)
> 	at edu.cmu.tetrad.search.Fges.scoreGraphChange(Fges.java:2027)
> 	at edu.cmu.tetrad.search.Fges.deleteEval(Fges.java:1525)
> 	at edu.cmu.tetrad.search.Fges.calculateArrowsBackward(Fges.java:1369)
> 	at edu.cmu.tetrad.search.Fges.initializeArrowsBackward(Fges.java:1106)
> 	at edu.cmu.tetrad.search.Fges.bes(Fges.java:1022)
> 	at edu.cmu.tetrad.search.Fges.search(Fges.java:265)
> 	at edu.cmu.tetrad.algcomparison.algorithm.oracle.pattern.Fges.search(Fges.java:90)
> 	at edu.pitt.dbmi.algo.resampling.task.GeneralResamplingSearchRunnable.run(GeneralResamplingSearchRunnable.java:133)
> 	at edu.pitt.dbmi.algo.resampling.GeneralResamplingSearch.search(GeneralResamplingSearch.java:360)
> 	at edu.pitt.dbmi.algo.resampling.GeneralResamplingTest.search(GeneralResamplingTest.java:199)
> 	at edu.cmu.tetrad.test.TestGeneralResamplingTest.testFGESc(TestGeneralResamplingTest.java:116)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
> 	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
> 	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:252)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:141)
> 	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:112)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
> 	at org.apache.maven.surefire.util.ReflectionUtils.invokeMethodWithArray(ReflectionUtils.java:189)
> 	at org.apache.maven.surefire.booter.ProviderFactory$ProviderProxy.invoke(ProviderFactory.java:165)
> 	at org.apache.maven.surefire.booter.ProviderFactory.invokeProvider(ProviderFactory.java:85)
> 	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:115)
> 	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:75)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1307#issuecomment-715937704>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACLFSR3ECC4VMKG5ZE7TA53SMLXIRANCNFSM4SYZGLOQ>
> .
>


-- 
Joseph D. Ramsey
Special Faculty and Director of Research Computing
Department of Philosophy
135 Baker Hall
Carnegie Mellon University
Pittsburgh, PA 15213

jsph.ramsey@gmail.com
Office: (412) 268-8063
http://www.andrew.cmu.edu/user/jdramsey

kvb2univpitt commented:
@jdramsey I was able to compile without errors now.  I am testing it now.

jdramsey commented:
Great!

On Sat, Oct 24, 2020 at 1:47 PM Kevin Bui <notifications@github.com> wrote:

> @jdramsey <https://github.com/jdramsey> I was able to compile without
> errors now. I am testing it now.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cmu-phil/tetrad/pull/1307#issuecomment-716030283>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACLFSR7K2PZK5H5OO44UXR3SMMHLBANCNFSM4SYZGLOQ>
> .
>


-- 
Joseph D. Ramsey
Special Faculty and Director of Research Computing
Department of Philosophy
135 Baker Hall
Carnegie Mellon University
Pittsburgh, PA 15213

jsph.ramsey@gmail.com
Office: (412) 268-8063
http://www.andrew.cmu.edu/user/jdramsey
